middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -208,150 +323,407 @@ let get_exn_extra_args env k =
+     must_inline_and_duplicate (since it basically replaces a variable by either
+     another variable, a constant, or a symbol). *)
+  match bound_expr with
+  | Simple { cmm_expr } when is_cmm_simple cmm_expr ->
+    create_binding_aux ?extra env effs var ~inline:Must_inline_and_duplicate
+      (Split { cmm_expr })
+  | Simple _ | Split _ | Splittable_prim _ ->
+    create_binding_aux ?extra env effs var ~inline bound_expr
+
+let bind_variable_with_decision (type a) ?extra env var ~inline
+    ~(defining_expr : a bound_expr) ~effects_and_coeffects_of_defining_expr:effs
+    =
+  (* CR gbury: we actually need to "lie" about the effects and coeffects of
+     allocations that we must inline, or else we end up with expressions with
+     effects that will not be inlined (see [to_cmm_primitive.ml] and in
+     particular ~consider_inllining_effectful_expressions). For instance,
+     consider: *)
+  (*
+   * let x = box_number 1. in
+   * let y = unbox_number x in
+   * let z = y +. 1. in
+   * ...
+   *)
+  (* because we only ever join effects when translating primitives, and that we
+     only use the effects and coeffects from the flambda code, we generate:
+
+     - a binding "x -> box_number 1." with generative effects (as expected)
+
+     - a binding "y -> 1." (because we inlined "x", and cmm_helper functions
+     eliminated the unbox-of-box), but this binding has generative effects,
+     because it effects are computed as : ece(prim:unbox_number) ∪ ece(x), and
+     since "x" has generative effects, we end up with a binding for "y" that has
+     generative effects.
+
+     - when we translate the addition, we explicitly do not consider inlining
+     effectful expressions, and thus we do not inline "y", since it has effects.
+
+     To counteract that, we instead consider that generative effects arising
+     from Delay/"must inline" primitives do not count, and that we consider the
+     binding to be pure.
+
+     CR gbury: this allows to move allocations marked as `Must_inline_once` past
+     function calls (and other effectful expressions), which can break some
+     allocation-counting tests (the same also applies to allocations marked as
+     `Must_inline_and_duplicate` but that is more expected). *)
+  let effs, classification =
+    let classification =
+      To_cmm_effects.classify_by_effects_and_coeffects effs
     in
-    env, binding
-
-let bind_variable0 ?extra env var ~effects_and_coeffects_of_defining_expr:effs
-    ~may_inline ~defining_expr =
-  let env, binding =
-    create_binding ?extra env ~may_inline effs var defining_expr
+    let change_eff_to effs =
+      effs, To_cmm_effects.classify_by_effects_and_coeffects effs
+    in
+    match[@ocaml.warning "-4"] (inline : a inline), classification with
+    | Must_inline_once, Generative_immutable -> change_eff_to Ece.pure
+    | Must_inline_and_duplicate, Generative_immutable ->
+      change_eff_to Ece.pure_can_be_duplicated
+    | _, _ -> effs, classification
```

Comment:
As we discussed earlier, I think this can be simplified to:
```ocaml
match inline, classification with
| (Must_inline_once | Must_inline_and_duplicate), Generative_immutable ->
  Ece.pure_can_be_duplicated, To_cmm_effects.Pure
| _, _ -> effs, classification
```

Reply:
changed (this required to remove the `private` on the type definition of `To_cmm_effects.classification` but it shouldn't be a problem).